### ISSUE TITLE
Problem with scope in translations:

### ISSUE THREAD

**Issue by diegomarguerettaz** [OrchardCMS/Orchard#8318]
If I want to translate the string "hello" in Italian and such string is contained in the class A, I add in the files . po the translation:
msgctxt "...A"
msgid "Hello"
msgstr "Ciao".
Now, if I have class B inheriting from A, when I try to call the translation from B, orchard can’t find the translation as it looks for it with scope B and not with scope A. It would be good thing if, when it doesn’t find the translation with scope B, try searching for such a translation with scope A. An example in Orchard.Framework can be found in Classes DefaultOrchardCommandHandler and FeatureCommands.

To make this fix you should engage:
- In the LocalizationModule, add to the dictionary of localizers both the one with the scope of the concrete class, and the one with the scope of the abstract class.
- From the concrete scope, one must be able to go back to the abstract one, perhaps saving the thing somewhere (additional dictionary of cache), or putting it directly in the objects of class Text.
- LocaliLocaliLocaliManager Manager Manager has parentalscope information to attempt a fallback if it is not located... "Final" fallbacks must remain as they are now (empty scope and then do not translate)

@MatteoPiovanelli-Laser

**Comment by MatteoPiovanelli** [comment]
**To test this issue**
 - Build the solution in debug mode.
 - Within Visual Studio, set a debug point on https://github.com/OrchardCMS/Orchard/blob/ffaa98724af14bc32860519f05b4a10383b28b2c/src/Orchard/Commands/DefaultOrchardCommandHandler.cs#L18
 - Set another debug point on https://github.com/OrchardCMS/Orchard/blob/ffaa98724af14bc32860519f05b4a10383b28b2c/src/Orchard.Web/Modules/Orchard.Modules/Commands/FeatureCommands.cs#L62
 - Open the command line utility Orchard.exe.
 - From Visual Studio, attach the debugger to the process for Orchard.exe.
 - In the command line tool, execute the command `feature enable Orchard.CodeGeneration`. You may use any feature name.
 - From both debug points, verify (watch) the scope that is set within the Localizer `T`.

**What we have**
In both cases the scope will be the same, ending with "FeatureCommands".

The net result of this issue is that to see the correct localization for the strings that are written in `DefaultOrchardCommandHandler.cs` we would have to add them to po files once for every concrete implementation of that abstract class.

**Likely cause**
The cause of this issue is that when the `Localizer` instances are created in `LocalizationModule` they are fed the scope that is computed based on the `LimitType` property, i.e. the actual class being instanced, with no reference whatsoever to any other class where the `Localizer` itself may be used. This is only correct in absence of inheritance.

**Proposed solution:**
 - Add to `ILocalizedStringManager` a method that accepts an ordered `IEnumerable` of strings (one for each scope). The implementation of this method would be similar to the one for `GetLocalizedString(string, string, string)` (https://github.com/OrchardCMS/Orchard/blob/ffaa98724af14bc32860519f05b4a10383b28b2c/src/Orchard/Localization/Services/DefaultLocalizedStringManager.cs#L56) except that it would check each scope in turn before moving on to the generic fallback. Similarly for the `GetParentTranslation(string, string, string)` that is already there.
 - Add to `Text` a constructor that accepts an ordered `IEnumerable` of strings (one for each scope).
 - Update the `Get(string, params object[])` method of `Text` so that it attempts to use all the scopes it has when it has several.
 - Add a method to `LocalizationUtilities` that accepts an ordered `IEnumerable` of strings, one for each scope, and that calls the new constructor of `Text`. This method would be similar to https://github.com/OrchardCMS/Orchard/blob/ffaa98724af14bc32860519f05b4a10383b28b2c/src/Orchard/Localization/LocalizationUtilities.cs#L15
 - In `LocalizationModule`, change the code starting https://github.com/OrchardCMS/Orchard/blob/ffaa98724af14bc32860519f05b4a10383b28b2c/src/Orchard/Localization/LocalizationModule.cs#L25 to be able to find/handle the multiple scopes.

Those additional scopes are not hard to determine in `LocalizationModule`, by recursively looking at the current type's `BaseType` property. (a good stop condition would likely be when the `BaseType` is `object`).

The breaking change to `ILocalizedStringManager` is probably required for this to work, because I can't think of a way to take the results of `GetParentTranslation(string, string, string)` from successive scopes and figure out which one is the right one.

**Comment by sebastienros** [comment]
Or add some hint like an attribute on the class to change the scope to something else?

**Comment by sebastienros** [comment]
Agreed on defining a list of scopes on the Localizer to fall back to in order.
